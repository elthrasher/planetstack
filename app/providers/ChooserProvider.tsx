import React, { createContext, ReactNode, useState } from 'react';

export enum ChooserMode {
  NONE,
  BACKGROUND,
  ICON,
  INSTRUCTIONS,
}

export interface ChooserContextModel {
  chooserMode: ChooserMode;
  setChooserMode: (chooserMode: ChooserMode) => void;
}

export const ChooserContext = createContext<ChooserContextModel>({
  chooserMode: ChooserMode.NONE,
  setChooserMode: /* istanbul ignore next */ () => null,
});

export const ChooserProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [chooserMode, setChooserMode] = useState<ChooserMode>(ChooserMode.NONE);

  return <ChooserContext.Provider value={{ chooserMode, setChooserMode }}>{children}</ChooserContext.Provider>;
};
