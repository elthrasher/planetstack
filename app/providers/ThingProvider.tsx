import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket from 'react-use-websocket';

const socketUrl = 'wss://jdnif5uxok.execute-api.us-east-1.amazonaws.com/dev';

export interface Thing {
  id: string;
  img: string;
  x: number;
  y: number;
}

export interface ThingContextModel {
  things: Thing[];
  addThing: (thing: Thing) => void;
  removeThing: (thing: Thing) => void;
}

export const ThingContext = createContext<ThingContextModel>({
  things: [],
  addThing: () => null,
  removeThing: () => null,
});

export const ThingProvider = ({ children }: { children: ReactNode }) => {
  const [things, setThings] = useState<Thing[]>([]);
  const addThing = (thing: Thing) => setThings([...things, thing]);
  const removeThing = (thing: Thing) => setThings(things.filter((t) => t !== thing));
  const { sendMessage, sendJsonMessage, lastMessage, lastJsonMessage, readyState, getWebSocket } = useWebSocket(
    socketUrl,
    {
      onError: (e) => console.error(e),
      onOpen: () => console.log(),
      //Will attempt to reconnect on all close events, such as server shutting down
      shouldReconnect: () => true,
    },
  );
  console.log(readyState);
  return <ThingContext.Provider value={{ things, addThing, removeThing }}>{children}</ThingContext.Provider>;
};
