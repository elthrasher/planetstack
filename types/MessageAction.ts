export enum MessageAction {
  ADD_ICON = 'addIcon',
  CHANGE_BACKGROUND = 'changeBackground',
  DELETE_ICON = 'deleteIcon',
  GET_STATE = 'getState',
  MOVE_ICON = 'moveIcon',
}
