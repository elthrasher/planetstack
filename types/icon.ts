export type IconModel = {
  img: number;
  x: number;
  y: number;
};
