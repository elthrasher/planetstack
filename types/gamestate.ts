import { IconModel } from './icon';

export type GameStateModel = {
  bg: number;
  icons: { [ksuid: string]: IconModel };
};
