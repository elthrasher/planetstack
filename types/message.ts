import { IconModel } from './icon';
import { MessageAction } from './MessageAction';

export interface AddIconModel {
  action: MessageAction.ADD_ICON;
  icon: IconModel;
}

export interface ChangeBackgroundModel {
  action: MessageAction.CHANGE_BACKGROUND;
  bg: number;
}

export interface DeleteIconModel {
  action: MessageAction.DELETE_ICON;
  id: string;
}

export interface GetStateModel {
  action: MessageAction.GET_STATE;
}

export interface MoveIconModel {
  action: MessageAction.MOVE_ICON;
  id: string;
  img: number;
  x: number;
  y: number;
}

export type MessageModel = AddIconModel | ChangeBackgroundModel | DeleteIconModel | GetStateModel | MoveIconModel;
