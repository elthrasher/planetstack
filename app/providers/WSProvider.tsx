import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket, { SendMessage } from 'react-use-websocket';

import { GameStateModel } from '../../types/gamestate';
import { MessageAction } from '../../types/MessageAction';

/* istanbul ignore next */
const getSocketUrl = async () => {
  const response = await fetch('./config.json');
  const config = await response.json();
  return `${config.socketUrl}/${config.stageName}`;
};

const defaultGs: GameStateModel = { bg: 0, icons: {} };

export interface WSContextModel {
  gs: GameStateModel;
  sendMessage: SendMessage;
}

export const WSContext = createContext<WSContextModel>({
  gs: defaultGs,
  sendMessage: () => null,
});

export const WSProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [gs, setGs] = useState(defaultGs);
  /* istanbul ignore next */
  const { sendMessage } = useWebSocket(getSocketUrl, {
    onOpen: () => sendMessage(JSON.stringify({ action: MessageAction.GET_STATE })),
    onError: (e) => console.error(e),
    onMessage: (message: MessageEvent<string>) => setGs(JSON.parse(message.data)),
    share: true,
    shouldReconnect: () => true,
  });
  return <WSContext.Provider value={{ gs, sendMessage }}>{children}</WSContext.Provider>;
};
