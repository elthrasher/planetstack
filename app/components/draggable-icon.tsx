import React, { MouseEvent, useContext, useMemo, useState } from 'react';
import Draggable, { DraggableData, DraggableEvent } from 'react-draggable';

import { icons } from '../../types/data/icons';
import { IconModel } from '../../types/icon';
import { MessageAction } from '../../types/MessageAction';
import { WSContext } from '../providers/WSProvider';

export const DraggableIcon = ({ icon, id }: { icon: IconModel; id: string }): JSX.Element => {
  const { sendJsonMessage } = useContext(WSContext);
  // Track local state for dragging, but update if we get a new message
  const [pos, setPos] = useState({ x: icon.x, y: icon.y });
  useMemo(() => setPos({ x: icon.x, y: icon.y }), [icon]);
  const handleContextMenu = (e: MouseEvent<HTMLImageElement>) => {
    e.preventDefault();
    sendJsonMessage({ action: MessageAction.DELETE_ICON, id });
  };
  const handleStop = (_e: DraggableEvent, d: DraggableData) => {
    setPos({ x: d.x, y: d.y });
    sendJsonMessage({ action: MessageAction.MOVE_ICON, id, img: icon.img, x: d.x, y: d.y });
  };
  return (
    <Draggable position={pos} onStop={handleStop}>
      <img className="thing" onContextMenu={handleContextMenu} src={icons[icon.img].path} />
    </Draggable>
  );
};
