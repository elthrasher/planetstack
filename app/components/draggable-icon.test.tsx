import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';

import { MessageAction } from '../../types/MessageAction';
import { WSContext } from '../providers/WSProvider';
import { DraggableIcon } from './draggable-icon';

const sendJsonMessage = jest.fn();

// Thanks https://github.com/testing-library/user-event/issues/440
const fireMouseEvent = function (type: string, elem: EventTarget, centerX: number, centerY: number) {
  act(() => {
    const evt = document.createEvent('MouseEvents');
    evt.initMouseEvent(type, true, true, window, 1, 1, 1, centerX, centerY, false, false, false, false, 0, elem);
    elem.dispatchEvent(evt);
  });
};

describe('draggable icon component', () => {
  test('Move from 400x400 to 499x499', () => {
    const { container, getByRole } = render(<DraggableIcon icon={{ img: 0, x: 400, y: 400 }} id={'abc-123'} />);
    expect(container).toMatchSnapshot();
    const img = getByRole('img');
    expect(img).toHaveClass('thing');
    fireMouseEvent('mousemove', img, 401, 401);
    fireMouseEvent('mouseenter', img, 401, 401);
    fireMouseEvent('mouseover', img, 401, 401);
    fireMouseEvent('mousedown', img, 401, 401);

    fireMouseEvent('drag', img, 401, 401);
    fireMouseEvent('mousemove', img, 401, 401);
    fireMouseEvent('drag', img, 500, 500);
    fireMouseEvent('mousemove', img, 500, 500);
    fireMouseEvent('drop', img, 500, 500);
    fireMouseEvent('dragend', img, 500, 500);
    fireMouseEvent('mouseup', img, 500, 500);
    expect(container).toMatchSnapshot();
    expect(img).toHaveAttribute('style', 'transform: translate(499px,499px);');
  });
  test('Remove the component', () => {
    const { container, getByRole } = render(
      <WSContext.Provider value={{ gs: { bg: 0, icons: {} }, sendJsonMessage }}>
        <DraggableIcon icon={{ img: 0, x: 400, y: 400 }} id={'abc-123'} />
      </WSContext.Provider>,
    );
    expect(container).toMatchSnapshot();
    const img = getByRole('img');
    expect(img).toHaveClass('thing');
    fireEvent.contextMenu(img);
    expect(sendJsonMessage).toHaveBeenCalledWith({ action: MessageAction.DELETE_ICON, id: 'abc-123' });
  });
});
