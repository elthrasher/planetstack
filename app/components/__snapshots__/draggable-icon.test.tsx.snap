// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draggable icon component Move from 400x400 to 499x499 1`] = `
<div>
  <img
    class="thing react-draggable"
    src="assets/RandomBuildings/B01.png"
    style="transform: translate(400px,400px);"
  />
</div>
`;

exports[`draggable icon component Move from 400x400 to 499x499 2`] = `
<div>
  <img
    class="thing react-draggable react-draggable-dragged"
    src="assets/RandomBuildings/B01.png"
    style="transform: translate(499px,499px);"
  />
</div>
`;

exports[`draggable icon component Remove the component 1`] = `
<div>
  <img
    class="thing react-draggable"
    src="assets/RandomBuildings/B01.png"
    style="transform: translate(400px,400px);"
  />
</div>
`;
